src/twine/decode.ml

```diff
@@ -4,19 +4,107 @@ module Slice = Byte_slice
 module LEB128 = Imandrakit_leb128.Decode
 
 type cached = ..
-type cached += Miss
 
 module type CACHE_KEY = sig
   type elt
   type cached += C of elt
 end
 
-type t = {
-  sl: slice;
-  cache: cached array;
-  mutable hmap: Hmap.t;
+type t =
+  | Decode : {
+      st: 'st;
+      ops: 'st ops;
+      cache: cached Int_tbl.t;
+      mutable hmap: Hmap.t;
+    }
+      -> t
+
+and 'st ops = {
+  length: 'st -> int;
+  read_char: 'st -> int -> char;
+  read_int32: 'st -> int -> int32;
+  read_int64: 'st -> int -> int64;
+  read_blob: 'st -> int -> int -> slice;
+  read_leb128: 'st -> int -> int64 * offset;
 }
 
+let st_len t =
+  let (Decode { st; ops; _ }) = t in
+  ops.length st
+
+let read_char t offset =
+  let (Decode { st; ops; _ }) = t in
+  ops.read_char st offset
+
+let read_int32 t offset =
+  let (Decode { st; ops; _ }) = t in
+  ops.read_int32 st offset
+
+let read_int64 t offset =
+  let (Decode { st; ops; _ }) = t in
+  ops.read_int64 st offset
+
+let read_blob t offset length =
+  let (Decode { st; ops; _ }) = t in
+  ops.read_blob st offset length
+
+let read_leb128 t offset =
+  let (Decode { st; ops; _ }) = t in
+  ops.read_leb128 st offset
+
+let slice_ops : slice ops =
+  {
+    length = (fun (st : slice) -> Slice.len st);
+    read_char = (fun (st : slice) (offset : int) -> Slice.get st offset);
+    read_int32 =
+      (fun (st : slice) (offset : int) ->
+        Bytes.get_int32_le st.bs (st.off + offset));
+    read_int64 =
+      (fun (st : slice) (offset : int) ->
+        Bytes.get_int64_le st.bs (st.off + offset));
+    read_blob =
+      (fun (st : slice) (offset : int) (length : int) ->
+        Slice.sub st offset (min length (Slice.len st - st.off - offset)));
+    read_leb128 = (fun (st : slice) (offset : int) -> LEB128.u64 st offset);
+  }
+
+let in_channel_ops : in_channel ops =
+  {
+    length = (fun (st : in_channel) -> Int64.to_int (In_channel.length st));
+    read_char =
+      (fun (st : in_channel) (offset : int) ->
+        In_channel.seek st (Int64.of_int offset);
+        Option.get (In_channel.input_char st));
+    read_int32 =
+      (fun (st : in_channel) (offset : int) ->
+        let bs = Bytes.create 4 in
+        In_channel.seek st (Int64.of_int offset);
+        let _ = In_channel.input st bs 0 4 in
```

Comment:
I think we need to use `really_input` instead here. A partial read would make us decode garbage on the next line

Reply:
Yep makes sense.